File: bindings/java/src/main/java/org/indunet/robot/bus/Context.java
package org.indunet.robot.bus;

import com.sun.jna.Pointer;

/**
 * Shared ZeroMQ runtime context. Same-process inproc between an embedded {@link Broker} and
 * {@link Node}s requires one shared {@code Context}.
 */
public final class Context implements AutoCloseable {
    private Pointer ptr;

    public Context() {
        this.ptr = Errors.checkPtr(RobotBusC.Holder.INSTANCE.robot_bus_context_new(), "Context");
    }

    private Context(Pointer ptr) {
        this.ptr = ptr;
    }

    /** Cheap clone (refcounted ZMQ context). */
    public Context cloneContext() {
        return new Context(
                Errors.checkPtr(RobotBusC.Holder.INSTANCE.robot_bus_context_clone(ptr), "Context.clone"));
    }

    Pointer raw() {
        return ptr;
    }

    @Override
    public void close() {
        if (ptr != null) {
            RobotBusC.Holder.INSTANCE.robot_bus_context_free(ptr);
            ptr = null;
        }
    }
}

File: bindings/android/src/main/kotlin/org/indunet/robot/bus/Context.kt
package org.indunet.robot.bus

import com.sun.jna.Pointer

/**
 * Shared ZeroMQ runtime context. Same-process inproc between an embedded [Broker] and
 * [Node]s requires one shared `Context`.
 */
class Context : AutoCloseable {
    private var ptr: Pointer?

    constructor() {
        ptr = Errors.checkPtr(RobotBusC.Holder.INSTANCE.robot_bus_context_new(), "Context")
    }

    private constructor(ptr: Pointer) {
        this.ptr = ptr
    }

    /** Cheap clone (refcounted ZMQ context). */
    fun cloneContext(): Context =
        Context(
            Errors.checkPtr(
                RobotBusC.Holder.INSTANCE.robot_bus_context_clone(ptr),
                "Context.clone",
            ),
        )

    internal fun raw(): Pointer? = ptr

    override fun close() {
        if (ptr != null) {
            RobotBusC.Holder.INSTANCE.robot_bus_context_free(ptr)
            ptr = null
        }
    }
}
